velox/functions/sparksql/aggregates/AverageAggregate.cpp

```diff
@@ -74,6 +74,319 @@ class AverageAggregate
+  explicit DecimalAverageAggregate(TypePtr resultType)
+      : DecimalAggregate<TInputType>(resultType) {}
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    decodedPartial_.decode(*args[0], rows);
+    auto baseRowVector = decodedPartial_.base()->template as<RowVector>();
+
+    if (this->validateIntermediateInputs_ &&
+        (baseRowVector->childAt(0)->mayHaveNulls() ||
+         baseRowVector->childAt(1)->mayHaveNulls())) {
+      addIntermediateResultsImpl<true>(groups, rows);
+      return;
+    }
+    addIntermediateResultsImpl<false>(groups, rows);
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /* mayPushdown */) override {
+    decodedPartial_.decode(*args[0], rows);
+    auto baseRowVector = decodedPartial_.base()->template as<RowVector>();
+
+    if (this->validateIntermediateInputs_ &&
+        (baseRowVector->childAt(0)->mayHaveNulls() ||
+         baseRowVector->childAt(1)->mayHaveNulls())) {
+      addSingleGroupIntermediateResultsImpl<true>(group, rows);
+      return;
+    }
+    addSingleGroupIntermediateResultsImpl<false>(group, rows);
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto rowVector = (*result)->as<RowVector>();
+    auto sumVector = rowVector->childAt(0)->asFlatVector<int128_t>();
+    auto countVector = rowVector->childAt(1)->asFlatVector<int64_t>();
+    rowVector->resize(numGroups);
+    sumVector->resize(numGroups);
+    countVector->resize(numGroups);
+    rowVector->clearAllNulls();
+
+    int64_t* rawCounts = countVector->mutableRawValues();
+    int128_t* rawSums = sumVector->mutableRawValues();
+    for (auto i = 0; i < numGroups; ++i) {
+      char* group = groups[i];
+      auto* accumulator = this->decimalAccumulator(group);
+      rawCounts[i] = accumulator->count;
+      rawSums[i] = accumulator->sum;
+    }
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto vector = (*result)->as<FlatVector<TResultType>>();
+    VELOX_CHECK(vector);
+    vector->resize(numGroups);
+    uint64_t* rawNulls = this->getRawNulls(vector);
+
+    TResultType* rawValues = vector->mutableRawValues();
+    for (int32_t i = 0; i < numGroups; ++i) {
+      char* group = groups[i];
+      auto accumulator = this->decimalAccumulator(group);
+      if (accumulator->count == 0) {
+        // In Spark, if all inputs are null, count will be 0,
+        // and the result of final avg will be null.
+        vector->setNull(i, true);
+      } else {
+        this->clearNull(rawNulls, i);
+        try {
+          rawValues[i] = computeAvg(accumulator);
+        } catch (const VeloxException& err) {
+          if (err.message().find("overflow") != std::string::npos ||
```

Comment:
I will change this part to overflow flag.

Reply:
Hi @mbasmanova , Spark SQL needs to return null when an overflow occurs, but the behavior of velox is to throw an exception, so we used to return a null value by catching the exception before. Do you have any other better suggestions for this part? For example, the framework level supports configuration whether to throw an exception or return a null value. cc @rui-mo

Reply:
@liujiayi771 Let's start by describing the semantics of Spark's avg function and how it differs from Presto. Make sure to include this in the PR description. Once we have that information, we can brainstorm on how best to implement it.

Reply:
> @liujiayi771 Let's start by describing the semantics of Spark's avg function and how it differs from Presto. Make sure to include this in the PR description. Once we have that information, we can brainstorm on how best to implement it.

The PR description has been updated.

Reply:
@mbasmanova When calculating the avg of decimal, `checkedMultiply` will be called. This method may overflow. Can we add a nullOnFailure parameter to these methods like the PR of try cast? The default is false, which will throw an exception.

Reply:
I will introduce `multiply` in this PR https://github.com/facebookincubator/velox/pull/4613/files#diff-4b21d26191cbbdb9bc90cafd83aed144a4483b1c594f224cf3bbe67f401f5f83R87

Reply:
> I will introduce `multiply` in this PR https://github.com/facebookincubator/velox/pull/4613/files#diff-4b21d26191cbbdb9bc90cafd83aed144a4483b1c594f224cf3bbe67f401f5f83R87

OK. I will use spark decimal multiply after this PR merged.